packages/core/echo/echo-pipeline/src/scheduler/timed-task-scheduler.ts

```diff
@@ -0,0 +1,174 @@
+//
+// Copyright 2025 DXOS.org
+//
+
+import { DeferredTask, sleepWithContext, asyncTimeout } from '@dxos/async';
+import { type Context, Resource } from '@dxos/context';
+
+const DEFAULT_TIMEOUT = 30_000;
+
+export type TimedTaskSchedulerParams = {
+  /**
+   * The time budget for the task execution.
+   * in [ms].
+   */
+  budget: number;
+
+  /**
+   * The time period for the budget calculation.
+   * in [ms].
+   */
+  budgetPeriod: number;
+
+  /**
+   * The rest time if the time budget is exceeded.
+   * in [ms].
+   */
+  restTime: number;
+
+  /**
+   * The maximum number of tasks to run in parallel.
+   */
+  maxParallelTasks: number;
+
+  /**
+   * Clean executed tasks after time period.
+   * in [ms].
+   */
+  cleanUpAfter: number;
+};
+
+export type Task<T> = {
+  run: () => Promise<T>;
+  result: Promise<T>;
+  resolve: (value: T) => void;
+  reject: (reason?: any) => void;
+
+  /**
+   * The start timestamp of the task.
+   */
+  start: number;
+
+  /**
+   * The end timestamp of the task.
+   */
+  end: number;
+
+  /**
+   * The elapsed time of the task.
+   * in [ms].
+   */
+  elapsed: number;
+
+  /**
+   * The timeout for the task.
+   * in [ms].
+   */
+  timeout: number;
+};
+/**
+ * Schedules and executes asynchronous tasks with time budgeting, parallelism, and rest periods.
+ * It is designed to enable backpressure for time-consuming tasks in a controlled manner.
+ */
+export class TimedTaskScheduler<T = void> extends Resource {
+  private readonly _executionQueue: Array<Task<T>> = [];
+  private _processedTasks: Array<Task<T>> = [];
+  private _executor?: DeferredTask = undefined;
+
+  constructor(private readonly _params: TimedTaskSchedulerParams) {
+    super();
+  }
+
+  /**
+   * Info method for debugging purposes.
+   */
+  get processedTasks(): Array<Task<T>> {
+    return this._processedTasks;
+  }
+
+  protected override async _open(ctx: Context): Promise<void> {
+    this._executor = new DeferredTask(ctx, async () => {
+      let now = Date.now();
+      // Clean up old tasks.
+      this._processedTasks = this._processedTasks.filter((task) => task.end > now - this._params.cleanUpAfter);
+
+      // Early return if there are no tasks to run.
+      if (this._executionQueue.length === 0) {
+        return;
+      }
+
+      // Check if the budget is exceeded.
+      if (this._processedTasks.length > 0) {
+        const lastPeriodTasks = this._processedTasks.filter(
```

Comment:
this will break down if you have multiple concurent tasks that are waiting for IO. You're gonna be counting the elapsed time for each one separatelly, even though they all executed concurently

Reply:
is there a repro of the hanging issue so we can compare before and after?